sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -275,36 +308,43 @@ class Dataset[T] private[sql](
     val numCols = schema.fieldNames.length
 
     // Initialise the width of each column to a minimum value of '3'
-    val colWidths = Array.fill(numCols)(3)
+    val colMaxWidths = Array.fill(numCols)(3)
+    val colWidths = Array.ofDim[Int](rows.length, numCols)
 
     // Compute the width of each column
+    var j = 0
     for (row <- rows) {
       for ((cell, i) <- row.zipWithIndex) {
-        colWidths(i) = math.max(colWidths(i), cell.length)
+        val width = unicodeWidth(cell)
+        colWidths(j)(i) = width
+        colMaxWidths(i) = math.max(colMaxWidths(i), width)
       }
+      j = j + 1
     }
 
     // Create SeparateLine
-    val sep: String = colWidths.map("-" * _).addString(sb, "+", "+", "+\n").toString()
+    val sep: String = colMaxWidths.map("-" * _).addString(sb, "+", "+", "+\n").toString()
 
     // column names
     rows.head.zipWithIndex.map { case (cell, i) =>
       if (truncate > 0) {
-        StringUtils.leftPad(cell, colWidths(i))
+        repeatPadding(colMaxWidths(i) - colWidths(0)(i)).append(cell)
```

Comment:
`StringUtils.leftPad/rightPad` uses String.length. Since this usage causes the same problem, the new code does not use these methods.